scripts/build_shapes.py

```diff
@@ -705,6 +726,259 @@ def _process_func_download_pop(c_code):
     )
 
 
+def get_worldpop_features(WorldPop_inputfile):
+    """
+    Function to extract data from .tif input file
+    -------
+    Inputs:
+        WorldPop_inputfile: String pointing to location of file
+    --------
+    Outputs:
+        src.meta["transform"]: Representation of the affine transform
+        src.shape: Dimensions of the input image
+    """
+    # Open the file using rasterio
+    with rasterio.open(WorldPop_inputfile) as src:
+        return src.meta["transform"], src.shape
+
+
+def get_worldpop_val_xy(WorldPop_inputfile, window_dimensions):
+    """
+    Function
+
+
+
+    """
+    col_off, row_off, width, height = window_dimensions
+
+    current_window = Window(col_off, row_off, width, height)
+
+    # Open the file using rasterio
+    with rasterio.open(WorldPop_inputfile) as src:
+        # --- Process the pixels in the image for population data ---
+        # Read the gray layer (1) to get an np.array of this band
+        # Rasterio doesn't support lower than float32 readout
+        # Hence np_pop_raster will have nbytes = 4 * shape[0] * shape[1]
+
+        np_pop_raster = src.read(1, window=current_window)
+
+        # Set np_pop_valid to all pixels in np_pop_raster that aren't 'nodata'
+        np_pop_valid = np_pop_raster[np_pop_raster != src.nodata]
+
+        # Set 'nodata' values to 0
+        np_pop_raster[np_pop_raster == src.nodata] = 0
+
+        # Set np_pop_xy to pixel locations of non zero values (same order as np_pop_valid)
+        np_pop_xy = np_pop_raster.nonzero()
+
+        # Transform to get [ x, y ] array
+        # np_pop_xy as 'I' (uintc), see
+        # https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.uintc
+        np_pop_xy = np.array([np_pop_xy[0], np_pop_xy[1]]).T.astype("I")
+
+    return np_pop_valid, np_pop_xy
+
+
+def compute_geomask_region(country_rows, affine_transform, dimensions):
+    """
+    Function
+
+
+
+    """
+    y_axis_len, x_axis_len = dimensions
+
+    # Set an empty numpy array with the dimensions of the country .tif file
+    # np_map_ID will contain a ID for each location (undefined is 0)
+    # ID corresponds to a specific geometry in country_rows
+    np_map_ID = np.zeros((y_axis_len, x_axis_len))
+
+    # List to contain the mappings of id to GADM_ID
+    id_to_GADM_ID = []
+
+    # Loop the country_rows geoDataFrame
+    for i in range(len(country_rows)):
```

Comment:
With this new revision we currently loose the paralleling capabilities that were available before.
If this new approach applied in large scale is really fast that may be ok, but my feeling is that we should preserve this feature for the final implementation

Reply:
Its possible to introduce parallel processing, however at a different stage then before, processing the countries in parallel rather then the image.

Reply:
This is a visualization of the .prof file for the previous run as shown in the comment above.

![build_shapes_world](https://user-images.githubusercontent.com/127969728/229588097-ffefde23-f5e4-4faf-9238-ad1b2cf886ab.svg)

The speed up is clearly more noticeable if you are working with a few countries vs a continent or the world.

Reply:
Parallel capabilities are re-implemented in add_population_data